packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart

```diff
@@ -0,0 +1,766 @@
+  final List<File> _inputs;
+  final List<File> _outputs;
+
+  /// Returns a [Depfile] that includes all input dependencies and non-temporary output
+  /// files.
+  Depfile get depfile => Depfile(_inputs, _outputs);
+
+  /// Returns true if there were any recommended changes that should
+  /// be applied.
+  ///
+  /// Retuns false if no problems were detected.
+  ///
+  /// If no checks are run, then this will default to false and will remain so
+  /// until a failing check finishes running.
+  bool get changesNeeded => _generatedFiles.isNotEmpty || _modifiedFiles.isNotEmpty
+    || (_goldenComparisonResults != null && !(_goldenComparisonResults['match'] as bool));
+
+  /// Checks if all dynamic feature modules are included in `settings.gradle`.
+  ///
+  /// `android/settings.gradle` should `include` all of the dynamic feature modules.
+  /// Each dynamic feature module holds the libs and asssets for a deferred component.
+  /// If the `settings.gradle` file does not have the proper includes, this method will
+  /// generate a modified settings.gradle with the correct include lines in the validator
+  /// output directory.
+  void checkAndroidSettingsGradle(List<DeferredComponent> components) {
+    // settings.gradle add ':componentName'
+    final File settingsGradle = env.projectDir.childDirectory('android').childFile('settings.gradle');
+    final File settingsGradleOutput = _outputDir.childFile('settings.gradle');
+    final List<String> lines = settingsGradle.readAsLinesSync();
+    if (settingsGradleOutput.existsSync()) {
+      settingsGradleOutput.deleteSync();
+    }
+    settingsGradleOutput.createSync(recursive: true);
+    // Parse out all included entries
+    final List<String> elements = <String>[];
+    for (final String line in lines) {
+      if (line.trim().startsWith('include')) {
+        elements.addAll(line.trim().substring(7).split(','));
+      }
+    }
+    // Clean and trim included entries
+    final List<String> trimmedElements = <String>[];
+    for (final String element in elements) {
+      trimmedElements.add(element.trim());
+    }
+    // Compute missing components
+    final List<String> missingComponents = <String>[];
+    for (final DeferredComponent component in components) {
+      final String componentName = '\':${component.name}\'';
+      if (!trimmedElements.contains(componentName)) {
+        missingComponents.add(componentName);
+      }
+    }
+    // Append missing components to first include line.
+    if (missingComponents.isNotEmpty) {
+      for (int i = 0; i < lines.length; i++) {
+        if (lines[i].trim().startsWith('include')) {
+          for (final String missingName in missingComponents) {
+            lines[i] += ', $missingName';
+          }
+          break;
+        }
+      }
+      // write lines.
+      for (final String line in lines) {
+        settingsGradleOutput.writeAsStringSync('$line\n', mode: FileMode.append, flush: true);
```

Comment:
prefer to write the file all at once, since it is not particularly big and to prevent cases where the tool getting killed would leave a partial file.